include/etl/span.h

```diff
@@ -710,15 +745,19 @@ namespace etl
     //*************************************************************************
     void advance(size_t elements) ETL_NOEXCEPT
     {
+      elements = etl::min(elements, size());
       pbegin += elements;
     }
 
     //*************************************************************************
     /// Reinterpret the span as a span with different element type.
     //*************************************************************************
     template<typename TNew>
-    ETL_NODISCARD ETL_CONSTEXPR etl::span<TNew, etl::dynamic_extent> reinterpret_as() const ETL_NOEXCEPT
+    ETL_NODISCARD ETL_CONSTEXPR14 etl::span<TNew, etl::dynamic_extent> reinterpret_as() const
     {
+#if ETL_USING_CPP11
```

Comment:
This does not check that the value of `pend - pbegin` is sized to a whole number of `TNew` elements.

If `pend - pbegin` is 3 bytes and `TNew` is an `int32_t` then the size of the new span is 0.

If `pend - pbegin` is 6 bytes and `TNew` is an `int32_t` then the size of the new span is 1, with 2 bytes before `pend`.

An advance of the `int32_t` `span` `pbegin` pointer would never equal `pend`, meaning loops of the return `span` would never terminate.